libraries/AP_InertialSensor/AP_InertialSensor_SCH16T.cpp

```diff
@@ -0,0 +1,472 @@
+#define SN_ID2          0x3E // 16 bit
+#define SN_ID3          0x3F // 16 bit
+
+#define T_STALL_US   20U
+
+#define SPI48_DATA_INT32(a)     (((int32_t)(((a) << 4)  & 0xfffff000UL)) >> 12)
+#define SPI48_DATA_UINT32(a)    ((uint32_t)(((a) >> 8)  & 0x000fffffUL))
+#define SPI48_DATA_UINT16(a)    ((uint16_t)(((a) >> 8)  & 0x0000ffffUL))
+
+extern const AP_HAL::HAL& hal;
+
+AP_InertialSensor_SCH16T::AP_InertialSensor_SCH16T(AP_InertialSensor &imu,
+                                                         AP_HAL::OwnPtr<AP_HAL::Device> _dev,
+                                                         enum Rotation _rotation,
+                                                         uint8_t drdy_gpio)
+    : AP_InertialSensor_Backend(imu)
+    , dev(std::move(_dev))
+    , rotation(_rotation)
+    , drdy_pin(drdy_gpio)
+{
+    expected_sample_rate_hz = 1475;
+    accel_scale = radians(1.f / 1600.f);
+    gyro_scale = radians(1.f / 1600.f);
+
+    _registers[0] = RegisterConfig(CTRL_FILT_RATE,  FILTER_68HZ);        // 68Hz -- default
+    _registers[1] = RegisterConfig(CTRL_FILT_ACC12, FILTER_68HZ);        // 68Hz -- default
+    _registers[2] = RegisterConfig(CTRL_FILT_ACC3,  FILTER_68HZ);        // 68Hz -- default
+    _registers[3] = RegisterConfig(CTRL_RATE,       RATE_300DPS_1475HZ); // +/- 300 deg/s, 1600 LSB/(deg/s) -- default, Decimation 8, 1475Hz
+    _registers[4] = RegisterConfig(CTRL_ACC12,      ACC12_8G_1475HZ);    // +/- 80 m/s^2, 3200 LSB/(m/s^2) -- default, Decimation 8, 1475Hz
+    _registers[5] = RegisterConfig(CTRL_ACC3,       ACC3_26G);           // +/- 260 m/s^2, 1600 LSB/(m/s^2) -- default
+}
+
+AP_InertialSensor_Backend *
+AP_InertialSensor_SCH16T::probe(AP_InertialSensor &imu,
+                                   AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                   enum Rotation rotation,
+                                   uint8_t drdy_gpio)
+{
+    if (!dev) {
+        return nullptr;
+    }
+    auto sensor = new AP_InertialSensor_SCH16T(imu, std::move(dev), rotation, drdy_gpio);
+
+    if (!sensor) {
+        return nullptr;
+    }
+
+    if (!sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+void AP_InertialSensor_SCH16T::start()
+{
+    hal.console->printf("AP_InertialSensor_SCH16T::start()");
+    if (!_imu.register_accel(accel_instance, expected_sample_rate_hz, dev->get_bus_id_devtype(DEVTYPE_INS_SCH16T)) ||
+        !_imu.register_gyro(gyro_instance, expected_sample_rate_hz,   dev->get_bus_id_devtype(DEVTYPE_INS_SCH16T))) {
+        return;
+    }
+
+    // setup sensor rotations from probe()
+    set_gyro_orientation(gyro_instance, rotation);
+    set_accel_orientation(accel_instance, rotation);
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_InertialSensor_SCH16T::loop, void),
+                                      "SCH16T",
+                                      1024, AP_HAL::Scheduler::PRIORITY_BOOST, 1)) {
```

Comment:
Quite a big stack, I doubt you need that much. Generally these are tied to the bus thread rather than creating a new separate thread.

Reply:
I'll move it to register_periodic_callback, I was using the adis1647 driver as a starting point